File: refs/a2a-latency-extension/samples/python/start.py
#!/usr/bin/env python3
import os
import sys
import subprocess
import yaml
from typing import List, Optional, TypedDict, Dict
import json
import argparse
import logging
logging.basicConfig(level=logging.INFO, format='%(asctime)s - %(name)s - %(levelname)s - %(message)s')
logger = logging.getLogger(__name__)

class ServerCmd(TypedDict):
    name: str
    command: str
    color: str
    env: Dict[str, str]
    silent: Optional[bool]


class SkillLatency(TypedDict):
    skill: str
    latency: int


class NgrokConfig(TypedDict):
    url: str


class ClientConfig(TypedDict):
    port: int
    supportsLatencyTaskUpdates: bool
    ngrok: Optional[NgrokConfig]
    system: str


class ServerConfig(TypedDict):
    name: str
    port: int
    supportsLatencyTaskUpdates: Optional[bool]
    skillLatency: Optional[List[SkillLatency]]


class Config(TypedDict):
    client: ClientConfig
    servers: List[ServerConfig]


CONFIG_FILE = 'config.yaml'


def get_config_file() -> str:
    """Parse command line arguments for --config flag."""
    parser = argparse.ArgumentParser(description='Start client and server processes')
    parser.add_argument('--config', type=str, default=CONFIG_FILE,
                       help=f'Config file path (default: {CONFIG_FILE})')
    args = parser.parse_args()
    return args.config


def main() -> None:
    config_file = get_config_file()

    # Check if config file exists
    if not os.path.exists(config_file):
        logger.error(f"Config file {config_file} not found!")
        sys.exit(1)

    try:
        # Read and parse the YAML configuration
        with open(config_file, 'r') as file:
            config = yaml.safe_load(file)

        if not config.get('servers') or len(config['servers']) == 0:
            logger.info('No servers configured in config.yaml')
            sys.exit(1)

        # Build commands for client and server
        commands: List[ServerCmd] = []

        # Client command (voice.py)
        client_cmd = "python ./src/client/voice.py"

        # Set environment variables for client configuration
        env = os.environ.copy()
        env['PYTHONPATH'] = os.getcwd()
        env['PORT'] = str(config['client']['port'])
        if config['client'].get('supportsLatencyTaskUpdates'):
            env['SUPPORTS_LATENCY_TASK_UPDATES'] = 'true'
        if config['client'].get('ngrok', {}).get('url'):
            env['NGROK_URL'] = config['client']['ngrok']['url']

        commands.append({
            'command': client_cmd,
            'name': 'Client',
            'env': env,
            'color': 'green',
            'silent': False
        })

        # Add commands for all servers
        server_colors = ['blue', 'cyan', 'magenta', 'yellow']
        for i, server in enumerate(config['servers']):
            server_cmd = "python ./src/server/app.py"

            # Set environment variables for the server
            server_env = os.environ.copy()
            server_env['PYTHONPATH'] = os.getcwd()
            server_env['PORT'] = str(server['port'])
            if server.get('supportsLatencyTaskUpdates'):
                server_env['SUPPORTS_LATENCY_TASK_UPDATES'] = 'true'
            if server.get('supportsExtension'):
                server_env['SUPPORTS_EXTENSION'] = 'true'
            if server.get('skillLatency'):
                server_env['SKILL_LATENCY'] = json.dumps(server['skillLatency'])

            commands.append({
                'command': server_cmd,
                'name': server['name'],
                'env': server_env,
                'color': server_colors[i % len(server_colors)],
                'silent': server.get('silent', False)
            })

        # Add ngrok command last if configured
        if config['client'].get('ngrok', {}).get('url'):
            # Use ngrok with explicit domain flag for better compatibility
            ngrok_cmd = f"ngrok http {config['client']['port']} --domain={config['client']['ngrok']['url']}"
            commands.append({
                'command': ngrok_cmd,
                'name': 'ngrok',
                'env': os.environ.copy(),
                'color': 'yellow',
                'silent': True
            })
            logger.info(f"🔗 Starting ngrok tunnel: {config['client']['ngrok']['url']} -> localhost:{config['client']['port']}")

        logger.info(f"Starting client and server{'s' if len(config['servers']) > 1 else ''}...")
        logger.info(f"Client port: {config['client']['port']}")
        for i, server in enumerate(config['servers']):
            logger.info(f"Server {i + 1} ({server['name']}) port: {server['port']}")

        if config['client'].get('ngrok', {}).get('url'):
            logger.info(f"ngrok URL: {config['client']['ngrok']['url']}")

        # Start all processes
        logger.info('Starting client and server processes...')
        processes: List[tuple[subprocess.Popen[bytes], str]] = []

        try:
            for cmd_info in commands:
                logger.info('Running command', cmd_info)

                # Configure output redirection based on silent flag
                stdout = subprocess.DEVNULL if cmd_info.get('silent', False) else None
                stderr = subprocess.DEVNULL if cmd_info.get('silent', False) else None

                process = subprocess.Popen(
                    cmd_info['command'],
                    shell=True,
                    env=cmd_info['env'],
                    stdout=stdout,
                    stderr=stderr
                )
                processes.append((process, cmd_info['name']))
                logger.info(f"Started {cmd_info['name']} process")

            # Wait for all processes
            logger.info("All processes started. Press Ctrl+C to terminate.")

            # Keep the main thread alive
            for process, name in processes:
                process.wait()

        except KeyboardInterrupt:
            logger.info('\nReceived Ctrl+C, shutting down gracefully...')
        finally:
            # Terminate all processes
            for process, name in processes:
                if process.poll() is None:  # If process is still running
                    logger.info(f"Terminating {name} process...")
                    process.terminate()
                    try:
                        process.wait(timeout=5)
                    except subprocess.TimeoutExpired:
                        logger.info(f"Killing {name} process...")
                        process.kill()

            logger.info("All processes terminated")

    except Exception as error:
        logger.error('Error reading or parsing config file:', error)
        sys.exit(1)


# Handle process termination
if __name__ == '__main__':
    try:
        main()
    except Exception as error:
        logger.error('Unexpected error:', error)
        sys.exit(1)

File: refs/a2a-latency-extension/samples/js/start.ts
#!/usr/bin/env tsx
import * as fs from 'fs';

import concurrently from 'concurrently';
import * as yaml from 'js-yaml';

interface ClientConfig {
  port: number;
  supportsLatencyTaskUpdates: boolean;
  ngrok?: {
    url: string;
  };
}

interface ServerConfig {
  name: string;
  port: number;
  supportsLatencyTaskUpdates?: boolean;
  supportsExtension?: boolean;
  skillLatency?: Array<{
    skill: string;
    latency: number;
  }>;
}

interface Config {
  client: ClientConfig;
  servers: ServerConfig[];
}

const CONFIG_FILE = 'config.yaml';

function getConfigFile(): string {
  // Parse command line arguments for --config flag
  const args = process.argv.slice(2);
  const configIndex = args.indexOf('--config');

  if (configIndex !== -1 && configIndex + 1 < args.length) {
    return args[configIndex + 1];
  }

  return CONFIG_FILE; // fallback to default
}

async function main() {
  const configFile = getConfigFile();

  // Check if config file exists
  if (!fs.existsSync(configFile)) {
    console.error(`Config file ${configFile} not found!`);
    process.exit(1);
  }

  try {
    // Read and parse the YAML configuration
    const configContent = fs.readFileSync(configFile, 'utf8');
    const config = yaml.load(configContent) as Config;

    if (!config.servers || config.servers.length === 0) {
      console.error('No servers configured in config.yaml');
      process.exit(1);
    }

    // Build commands for client and server
    const commands: Array<{
      command: string;
      name: string;
      prefixColor: string;
    }> = [];

    // Client command (voice.ts) - only start one instance
    let clientCmd = `tsx ./src/client/voice.ts`;

    // Set environment variables for the client
    let clientEnvVars = `PORT=${config.client.port} `;
    if (config.client.supportsLatencyTaskUpdates !== undefined) {
      clientEnvVars += `SUPPORTS_LATENCY_TASK_UPDATES=${config.client.supportsLatencyTaskUpdates} `;
    }
    if (config.client.ngrok?.url) {
      clientEnvVars += `NGROK_URL=${config.client.ngrok.url} `;
    }

    clientCmd = `${clientEnvVars}${clientCmd}`;

    commands.push({
      command: clientCmd,
      name: 'Client',
      prefixColor: 'green',
    });

    // Add commands for all servers
    const serverColorModifiers = [
      'blue',
      'blue.bold',
      'blue.dim',
      'blue.italic',
    ];
    config.servers.forEach((server, index) => {
      const serverCmd = `tsx ./src/server/index.ts`;
      // Set environment variables for the server
      let serverCmdWithEnv = `PORT=${server.port}`;
      if (server.supportsLatencyTaskUpdates !== undefined) {
        serverCmdWithEnv += ` SUPPORTS_LATENCY_TASK_UPDATES=${server.supportsLatencyTaskUpdates}`;
      }
      if (server.supportsExtension !== undefined) {
        serverCmdWithEnv += ` SUPPORTS_EXTENSION=${server.supportsExtension}`;
      }
      // Pass skillLatency configuration as JSON string
      if (server.skillLatency && server.skillLatency.length > 0) {
        const skillLatencyJson = JSON.stringify(server.skillLatency);
        serverCmdWithEnv += ` SKILL_LATENCY='${skillLatencyJson}'`;
      }
      serverCmdWithEnv += ` ${serverCmd}`;

      commands.push({
        command: serverCmdWithEnv,
        name: server.name,
        prefixColor: serverColorModifiers[index % serverColorModifiers.length],
      });
    });

    // Add ngrok command last if configured
    if (config.client.ngrok?.url) {
      // Use ngrok with explicit domain flag for better compatibility
      const ngrokCmd = `ngrok http ${config.client.port} --domain=${config.client.ngrok.url}`;
      commands.push({
        command: ngrokCmd,
        name: 'ngrok',
        prefixColor: 'yellow',
      });
      console.log(
        `🔗 Starting ngrok tunnel: ${config.client.ngrok.url} -> localhost:${config.client.port}`,
      );
    } else if (config.client.ngrok?.url) {
      console.log(
        `🔗 ngrok configured but disabled. To enable, set client.ngrok.enabled: true in config.yaml`,
      );
    }

    console.log(
      `Starting client and server${config.servers.length > 1 ? 's' : ''}...`,
    );
    console.log(`Client port: ${config.client.port}`);
    config.servers.forEach((server, index) => {
      console.log(`Server ${index + 1} (${server.name}) port: ${server.port}`);
    });
    if (config.client.ngrok?.url) {
      console.log(`ngrok URL: ${config.client.ngrok.url}`);
    }

    console.log('Starting client and server with concurrently...'); // Use concurrently API directly
    try {
      const prefixColors = [
        'green',
        'blue',
        'cyan',
        'magenta',
        'yellow',
        'red',
      ];
      const { result } = concurrently(commands, {
        prefix: 'name',
        killOthers: ['failure', 'success'],
        restartTries: 0,
        prefixColors: prefixColors.slice(0, commands.length),
      });

      // Wait for all processes to complete
      await result;
      console.log('Client and server completed successfully');
    } catch (error) {
      console.error('Error running client and server:');

      // Check if it's a concurrently error with process details
      if (Array.isArray(error) && error.length > 0) {
        error.forEach((processError, index) => {
          if (processError.command && processError.command.name) {
            console.error(
              `  ${processError.command.name} (exit code: ${processError.exitCode})`,
            );
            if (processError.command.name === 'ngrok') {
              console.error('  💡 ngrok failed - this might be due to:');
              console.error(
                "     - Authentication issues (run 'ngrok authtoken YOUR_TOKEN')",
              );
              console.error('     - Domain not available or not authorized');
              console.error('     - Network connectivity issues');
              console.error('     - Incorrect ngrok command syntax');
              console.error(
                "  💡 To disable ngrok, add 'enabled: false' to client.ngrok in config.yaml",
              );
            }
          }
        });
      } else {
        console.error(error);
      }
      process.exit(1);
    }
  } catch (error) {
    console.error('Error reading or parsing config file:', error);
    process.exit(1);
  }
}

// Handle process termination
process.on('SIGINT', () => {
  console.log('\nReceived SIGINT, shutting down gracefully...');
  process.exit(0);
});

process.on('SIGTERM', () => {
  console.log('\nReceived SIGTERM, shutting down gracefully...');
  process.exit(0);
});

main().catch((error) => {
  console.error('Unexpected error:', error);
  process.exit(1);
});
